File: frontend/src/api/documents.ts
import apiClient from './client';

// Документ = (карточка сметы, тип документа). Один контракт на все типы;
// чем типы отличаются — описывает row_format и набор разрешённых действий.
export type DocumentKind =
  | 'list' | 'completeness' | 'estimate' | 'optimization'
  // Раздел сводной: строки лежат снимком внутри сводной, версий у него нет.
  | 'summary-section';
export type RowFormat = 'generic' | 'estimate';
export type ReadonlyReason = 'no_permission' | 'task_processing' | 'input_readonly';

export interface DocumentRef {
  cardId: string;
  kind: DocumentKind;
  /** 'input' открывает документ исходного файла заказчика (только чтение). */
  fileSlot?: string;
  /** Номер входного файла, когда их несколько. */
  fileIndex?: number;
  versionId?: string;
}

export interface VersionBrief {
  id: string;
  version_number: number;
  version_label: string;
  version_display_name: string;
  is_rolled_back: boolean;
  created_at: string;
  // Проценты доп. расходов у каждой версии свои — сравнение версий считает
  // итоги по ним, а не по общей ставке проекта.
  overhead_pct: number;
  transport_pct: number;
  contingency_pct: number;
  expenses_overridden: boolean;
}

export interface LockInfo {
  user_id: number | null;
  user_name: string;
  heartbeat_at: string;
}

export interface DocumentMeta {
  card_id: string;
  kind: DocumentKind;
  row_format: RowFormat;
  file_slot: string;
  task_id: string;
  task_type: string;
  task_status: string;
  can_write: boolean;
  readonly_reason: ReadonlyReason | null;
  rev: number;
  active_version_id: string | null;
  versions: VersionBrief[];
  coefficient: Record<string, unknown> | null;
  has_draft: boolean;
  draft_updated_at: string | null;
  lock: LockInfo | null;
  project: { overhead_pct: number; transport_pct: number; name?: string };
  /** Только у раздела сводной: его строки разошлись со сметой. */
  divergence?: SectionDivergence | null;
}

/** Раздел сводной и смета показывают разное — обе стороны в цифрах. */
export interface SectionDivergence {
  section_rows: number;
  estimate_rows: number;
  section_total: number;
  estimate_total: number;
}

export interface DocumentRows {
  version_id: string;
  rev: number;
  rows: unknown[];
  draft_rows: unknown[] | null;
}

export interface ApplyResult {
  version_id: string;
  rev: number;
  rows_count: number;
  changes_count: number;
}

export interface HistoryChange {
  row_number: number;
  row_id: string | null;
  row_name: string;
  field: string;
  previous: unknown;
  new: unknown;
}

export interface HistoryEntry {
  id: string;
  kind: string | null;
  operation_type: string;
  description: string;
  user_id: number | null;
  user_name: string;
  created_at: string;
  changes_count: number;
  changes: HistoryChange[];
}

function base(ref: DocumentRef): string {
  return `/documents/${ref.cardId}/${ref.kind}`;
}

function slotParams(ref: DocumentRef): Record<string, string | number> {
  return {
    ...(ref.fileSlot ? { file_slot: ref.fileSlot } : {}),
    ...(ref.fileIndex !== undefined ? { file_index: ref.fileIndex } : {}),
  };
}

export async function getDocumentMeta(ref: DocumentRef): Promise<DocumentMeta> {
  const res = await apiClient.get<DocumentMeta>(base(ref), { params: slotParams(ref) });
  return res.data;
}

/**
 * Свести разошедшиеся раздел сводной и смету к одной стороне.
 *
 * `section` — верны правки раздела, они уезжают в смету; `estimate` — верна
 * смета, раздел берёт её строки. Прежние строки раздела в обоих случаях уходят
 * в историю.
 */
export async function resolveSectionDivergence(
  cardId: string,
  prefer: 'section' | 'estimate',
): Promise<{ prefer: string; rows_count: number }> {
  const res = await apiClient.post(
    `/documents/${cardId}/summary-section/divergence/resolve`, { prefer },
  );
  return res.data;
}

export async function getDocumentRows(ref: DocumentRef): Promise<DocumentRows> {
  const res = await apiClient.get<DocumentRows>(`${base(ref)}/rows`, {
    params: { ...slotParams(ref), ...(ref.versionId ? { version_id: ref.versionId } : {}) },
  });
  return res.data;
}

export async function saveDraft(
  ref: DocumentRef,
  versionId: string,
  rows: unknown[],
): Promise<void> {
  await apiClient.put(
    `${base(ref)}/draft`,
    { version_id: versionId, rows },
    { params: slotParams(ref) },
  );
}

export async function discardDraft(ref: DocumentRef, versionId: string): Promise<void> {
  await apiClient.delete(`${base(ref)}/draft`, {
    params: { ...slotParams(ref), version_id: versionId },
  });
}

export async function applyDocument(
  ref: DocumentRef,
  versionId: string,
  rev: number,
  rows?: unknown[],
): Promise<ApplyResult> {
  const res = await apiClient.post<ApplyResult>(
    `${base(ref)}/apply`,
    { version_id: versionId, rev, ...(rows ? { rows } : {}) },
    { params: slotParams(ref) },
  );
  return res.data;
}

export interface CoefficientPayload {
  work: number;
  material: number;
  /** 'all' — весь документ; список ключей строк — только отмеченные. */
  scope: 'all' | string[];
}

/** Поставить коэффициент к ценам или снять его (null). */
export async function setDocumentCoefficient(
  ref: DocumentRef,
  payload: CoefficientPayload | null,
): Promise<{ coefficient: CoefficientPayload | null }> {
  const res = await apiClient.put<{ coefficient: CoefficientPayload | null }>(
    `${base(ref)}/coefficient`, payload, { params: slotParams(ref) },
  );
  return res.data;
}

/** Выгрузка-ведомость по документу: строки приходят из предпросмотра. */
export async function exportDocument(
  ref: DocumentRef,
  payload: unknown,
  fileName = 'export.xlsx',
): Promise<void> {
  const res = await apiClient.post(`${base(ref)}/export`, payload, {
    params: slotParams(ref), responseType: 'blob',
  });
  const url = window.URL.createObjectURL(new Blob([res.data]));
  const link = document.createElement('a');
  link.href = url;
  link.setAttribute('download', fileName);
  document.body.appendChild(link);
  link.click();
  link.remove();
  window.URL.revokeObjectURL(url);
}

export async function getDocumentHistory(ref: DocumentRef): Promise<HistoryEntry[]> {
  const res = await apiClient.get<HistoryEntry[]>(`${base(ref)}/history`, {
    params: slotParams(ref),
  });
  return res.data;
}

export async function revertDocument(
  ref: DocumentRef,
  entryId: string,
  versionId?: string,
): Promise<ApplyResult> {
  const res = await apiClient.post<ApplyResult>(
    `${base(ref)}/history/${entryId}/revert`,
    null,
    { params: { ...slotParams(ref), ...(versionId ? { version_id: versionId } : {}) } },
  );
  return res.data;
}

export async function sendHeartbeat(ref: DocumentRef): Promise<LockInfo | null> {
  const res = await apiClient.post<{ lock: LockInfo | null }>(
    `${base(ref)}/heartbeat`,
    null,
    { params: slotParams(ref) },
  );
  return res.data.lock;
}

// --- Прайс ------------------------------------------------------------------

export interface PriceListItem {
  kind: 'work' | 'material';
  name: string;
  unit: string | null;
  price: number | null;
}

export interface PriceListSummary {
  added: number;
  updated: number;
  skipped: number;
  /** Почему позиции пропущены: «без цены» → 3. */
  skipped_reasons: Record<string, number>;
}

/**
 * Отправить позиции документа в общий прайс. Работы уходят к псевдо-подрядчику
 * «Из смет», материалы — ценой; сам документ при этом не меняется.
 */
export async function addToPriceList(
  ref: DocumentRef,
  items: PriceListItem[],
): Promise<PriceListSummary> {
  const res = await apiClient.post<PriceListSummary>(
    `${base(ref)}/price-list`,
    { items },
    { params: slotParams(ref) },
  );
  return res.data;
}

// --- Проверка единиц измерения у цен ----------------------------------------

export interface PriceUnitsCheckResult {
  /** Сколько строк с ценой проверено. */
  checked: number;
  /** В скольких единица цены разошлась с единицей позиции. */
  flagged: number;
  version_id: string;
  rev: number;
}

/**
 * Пройти по смете и пометить строки, где цена похожа на цену за другую единицу
 * измерения. Нужна сметам, посчитанным до того, как подбор цены начал сверять
 * единицу: цена за тонну могла встать в строку с килограммами.
 *
 * Цены не меняются — проверка показывает, где смотреть, решает человек.
 */
export async function checkPriceUnits(
  ref: DocumentRef,
  versionId: string,
  rev: number,
): Promise<PriceUnitsCheckResult> {
  const res = await apiClient.post<PriceUnitsCheckResult>(
    `${base(ref)}/price-units-check`,
    { version_id: versionId, rev },
    { params: slotParams(ref) },
  );
  return res.data;
}

// --- Поиск аналогов через ИИ -------------------------------------------------

export interface AnalogVariant {
  name: string;
  unit: string;
  price: number;
  /** Выгода в рублях по объёму позиции — считает сервер. */
  delta: number;
  reason: string;
  source: string;
}

export interface AnalogResult {
  row_id: string;
  name: string;
  unit: string;
  price: number;
  variants: AnalogVariant[];
}

export type AnalogStatus = 'queued' | 'running' | 'done' | 'failed' | 'cancelled';

export interface AnalogsState {
  run_id: string | null;
  status: AnalogStatus | null;
  processed: number;
  total: number;
  results: AnalogResult[];
  error: string | null;
  created_at: string | null;
}

export interface AnalogRowIn {
  row_id: string;
  name: string;
  unit: string | null;
  qty: number | null;
  price: number | null;
  kind: 'work' | 'material';
}

export interface AnalogsStartResult {
  run_id: string;
  status: AnalogStatus;
  total: number;
  estimate: { positions: number; searches: number; minutes: number };
}

/** Запустить фоновый поиск аналогов. Документ при этом не меняется. */
export async function startAnalogs(
  ref: DocumentRef,
  rows: AnalogRowIn[],
  versionId?: string,
): Promise<AnalogsStartResult> {
  const res = await apiClient.post<AnalogsStartResult>(
    `${base(ref)}/analogs`,
    { rows, version_id: versionId ?? null },
    { params: slotParams(ref) },
  );
  return res.data;
}

export async function getAnalogsState(ref: DocumentRef): Promise<AnalogsState> {
  const res = await apiClient.get<AnalogsState>(`${base(ref)}/analogs`, {
    params: slotParams(ref),
  });
  return res.data;
}

export async function cancelAnalogs(ref: DocumentRef): Promise<AnalogsState> {
  const res = await apiClient.post<AnalogsState>(`${base(ref)}/analogs/cancel`, null, {
    params: slotParams(ref),
  });
  return res.data;
}

/** Задача → документ. Для старых ссылок вида /tasks/{id}, где карточка неизвестна. */
export interface DocumentLocation {
  project_id: string;
  card_id: string;
  kind: DocumentKind;
}

export async function locateDocumentByTask(taskId: string): Promise<DocumentLocation> {
  const res = await apiClient.get<DocumentLocation>(`/documents/by-task/${taskId}`);
  return res.data;
}

/** Тип задачи → тип документа. Точка входа в редактор знает тип задачи. */
export function kindFromTaskType(taskType: string): DocumentKind | null {
  switch (taskType) {
    case 'LIST_FROM_GRAND':
    case 'LIST_FROM_PROJECT':
      return 'list';
    case 'CHECK_LIST_COMPLETENESS':
    case 'CHECK_PROJECT_COMPLETENESS':
      return 'completeness';
    case 'ESTIMATE_FROM_LIST':
      return 'estimate';
    case 'ESTIMATE_OPTIMIZATION':
      return 'optimization';
    default:
      return null;
  }
}

File: backend/app/routers/documents.py
"""Единый API документов — то, с чем работает единый редактор таблиц.

Один контракт на все типы документов карточки сметы. Вся логика прав, черновиков,
истории и присутствия живёт в `services/document_service.py`; здесь — только
HTTP-обвязка.
"""
import asyncio
import io
from typing import Optional
from urllib.parse import quote

import structlog
from fastapi import APIRouter, Body, Depends, HTTPException, Query, status
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas.document import (
    AnalogsStartRequest,
    AnalogsStartResponse,
    AnalogsStateResponse,
    ApplyRequest,
    ApplyResponse,
    CoefficientRequest,
    ExportRequest,
    DocumentMeta,
    DocumentRows,
    HeartbeatResponse,
    HistoryEntryOut,
    PriceListRequest,
    PriceListResponse,
    ProjectSettings,
    ResolveDivergenceRequest,
    SaveDraftRequest,
    SectionDivergence,
    VersionBrief,
)
from app.services import analogs_service
from app.services import document_service as svc
from app.utils.auth import get_current_user
from app.utils.xlsx_statement import generate_statement_xlsx

logger = structlog.get_logger()

router = APIRouter(prefix="/documents", tags=["documents"])

# Слот файла: 'input' открывает документ исходного файла заказчика (только чтение).
FileSlotQuery = Query(default=None, description="Слот файла: 'input' — исходный файл")
FileIndexQuery = Query(default=None, description="Номер входного файла (для file_slot='input')")


@router.get("/by-task/{task_id}")
async def locate_document_by_task(
    task_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Какой документ соответствует задаче — для старых ссылок вида /tasks/{id}."""
    project_id, card_id, kind = await svc.locate_by_task(db, task_id, current_user)
    return {"project_id": project_id, "card_id": card_id, "kind": kind}


@router.get("/{card_id}/{kind}", response_model=DocumentMeta)
async def get_document_meta(
    card_id: str,
    kind: str,
    file_slot: Optional[str] = FileSlotQuery,
    file_index: Optional[int] = FileIndexQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Метаданные документа: версии, права, черновик, коэффициент, присутствие."""
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot, file_index)
    # Первое открытие документа создаёт V0 из файла — раньше это делал клиент.
    doc = await svc.ensure_versions(db, doc, current_user, file_index)
    can_write, reason = svc.write_state(doc, current_user)
    active = doc.active
    lock = await svc.get_foreign_lock(db, card_id, kind, current_user)
    divergence = await svc.summary_divergence(db, doc)

    return DocumentMeta(
        card_id=str(doc.card.id),
        kind=doc.kind,
        row_format=doc.row_format,
        file_slot=doc.file_slot,
        task_id=str(doc.task.id),
        task_type=doc.task.task_type,
        task_status=doc.task.status,
        can_write=can_write,
        readonly_reason=reason,
        rev=active.rev if active else 0,
        active_version_id=str(active.id) if active else None,
        versions=[
            VersionBrief(
                id=str(v.id),
                version_number=v.version_number,
                version_label=v.version_label,
                version_display_name=v.version_display_name,
                is_rolled_back=v.is_rolled_back,
                created_at=v.created_at.isoformat(),
                overhead_pct=float(v.overhead_pct or 0),
                transport_pct=float(v.transport_pct or 0),
                contingency_pct=float(v.contingency_pct or 0),
                expenses_overridden=bool(v.expenses_overridden),
            )
            for v in doc.versions
        ],
        coefficient=active.coefficient if active else None,
        has_draft=bool(active and active.draft_rows is not None),
        draft_updated_at=(
            active.draft_updated_at.isoformat()
            if active and active.draft_updated_at else None
        ),
        lock=lock,
        project=ProjectSettings(
            overhead_pct=doc.project.overhead_pct,
            transport_pct=doc.project.transport_pct,
            name=doc.project.name or "",
        ),
        divergence=SectionDivergence(**divergence) if divergence else None,
    )


@router.post("/{card_id}/summary-section/divergence/resolve")
async def resolve_section_divergence(
    card_id: str,
    body: ResolveDivergenceRequest,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Свести разошедшиеся раздел и смету к одной стороне — по решению человека."""
    doc = await svc.resolve_document(db, card_id, "summary-section", current_user)
    return await svc.resolve_summary_divergence(db, doc, body.prefer, current_user)


@router.get("/{card_id}/{kind}/rows", response_model=DocumentRows)
async def get_document_rows(
    card_id: str,
    kind: str,
    version_id: Optional[str] = Query(default=None),
    file_slot: Optional[str] = FileSlotQuery,
    file_index: Optional[int] = FileIndexQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot, file_index)
    doc = await svc.ensure_versions(db, doc, current_user, file_index)
    version = svc.pick_version(doc, version_id)
    return DocumentRows(
        version_id=str(version.id),
        rev=version.rev or 0,
        # Строки берёт сервис: у раздела сводной они лежат не в версии, а
        # снимком внутри самой сводной.
        rows=svc.read_rows(doc, version),
        draft_rows=version.draft_rows,
    )


@router.put("/{card_id}/{kind}/draft")
async def save_document_draft(
    card_id: str,
    kind: str,
    body: SaveDraftRequest,
    file_slot: Optional[str] = FileSlotQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Автосохранение правок. Рабочие строки и файл не трогает."""
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot)
    version = svc.pick_version(doc, body.version_id)
    await svc.save_draft(db, doc, version, body.rows, current_user)
    return {"version_id": str(version.id), "rows_count": len(body.rows)}


@router.delete("/{card_id}/{kind}/draft")
async def drop_document_draft(
    card_id: str,
    kind: str,
    version_id: Optional[str] = Query(default=None),
    file_slot: Optional[str] = FileSlotQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Отказаться от непринятых правок."""
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot)
    version = svc.pick_version(doc, version_id)
    svc.ensure_writable(doc, current_user)
    await svc.discard_draft(db, version)
    await db.commit()
    return {"version_id": str(version.id), "status": "discarded"}


@router.post("/{card_id}/{kind}/apply", response_model=ApplyResponse)
async def apply_document(
    card_id: str,
    kind: str,
    body: ApplyRequest,
    file_slot: Optional[str] = FileSlotQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """«Применить»: черновик (или переданные строки) → рабочие строки."""
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot)
    version = svc.pick_version(doc, body.version_id)
    result = await svc.apply_rows(
        db, doc, version, body.rows, body.rev, current_user
    )
    return ApplyResponse(**result)


@router.put("/{card_id}/{kind}/coefficient")
async def set_document_coefficient(
    card_id: str,
    kind: str,
    body: Optional[CoefficientRequest] = Body(default=None),
    file_slot: Optional[str] = FileSlotQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Поставить коэффициент к ценам или снять его (пустое тело = снять).

    Исходные цены строк не меняются: коэффициент — настройка документа, его
    можно снять и получить прежние числа.
    """
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot)
    version = svc.pick_version(doc, None)
    payload = None
    if body is not None:
        scope = body.scope if body.scope == "all" else [str(x) for x in (body.scope or [])]
        payload = {"work": body.work, "material": body.material, "scope": scope}
    return await svc.set_coefficient(db, doc, version, payload, current_user)


@router.post("/{card_id}/{kind}/export")
async def export_document_statement(
    card_id: str,
    kind: str,
    body: ExportRequest,
    file_slot: Optional[str] = FileSlotQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Выгрузка-ведомость по документу.

    Строки приходят из предпросмотра — там человек их правит и удаляет, — а
    цены в них уже с коэффициентом: его применяет редактор при показе. Права
    проверяются на чтение: выгрузка ничего не меняет, поэтому доступна и там,
    где документ править нельзя (идёт расчёт, чужая смета у руководителя).
    """
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot)
    if not body.rows:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Нечего выгружать: не осталось ни одной строки",
        )
    if not body.columns:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Выберите хотя бы один столбец",
        )

    xlsx_bytes = await asyncio.to_thread(
        generate_statement_xlsx,
        [column.model_dump() for column in body.columns],
        body.rows,
        title=body.header.title,
        object_name=body.header.object_name,
        project_name=body.header.project_name or doc.project.name,
        show_date=body.header.show_date,
        show_total=body.header.show_total,
        sheet_name=body.sheet_name,
    )

    ascii_name = "statement.xlsx"
    utf8_name = quote(body.file_name or f"{body.header.title or 'Выгрузка'}.xlsx", safe="")
    return StreamingResponse(
        io.BytesIO(xlsx_bytes),
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={
            "Content-Disposition":
                f"attachment; filename=\"{ascii_name}\"; filename*=UTF-8''{utf8_name}",
        },
    )


@router.post("/{card_id}/{kind}/price-list", response_model=PriceListResponse)
async def add_document_rows_to_price_list(
    card_id: str,
    kind: str,
    body: PriceListRequest,
    file_slot: Optional[str] = FileSlotQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Отправить выделенные позиции документа в общий прайс.

    Работы уходят к псевдо-подрядчику «Из смет», материалы — ценой. Документ при
    этом не меняется: права проверяются на чтение, как у выгрузки.
    """
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot)
    items = [item.model_dump() for item in body.items]
    return await svc.add_to_price_list(db, doc, items, current_user)


@router.post("/{card_id}/{kind}/analogs", response_model=AnalogsStartResponse)
async def start_analogs_search(
    card_id: str,
    kind: str,
    body: AnalogsStartRequest,
    file_slot: Optional[str] = FileSlotQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Запустить поиск более дешёвых аналогов по выделенным позициям.

    Работа фоновая: поиск идёт в интернете и занимает минуты. Документ она не
    меняет — найденное человек принимает вручную.
    """
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot)
    run = await svc.start_analogs(
        db, doc, body.version_id, [row.model_dump() for row in body.rows], current_user,
    )
    return AnalogsStartResponse(
        run_id=str(run.id),
        status=run.status,
        total=run.total,
        estimate=analogs_service.estimate_effort(run.total),
    )


@router.get("/{card_id}/{kind}/analogs", response_model=AnalogsStateResponse)
async def get_analogs_state(
    card_id: str,
    kind: str,
    file_slot: Optional[str] = FileSlotQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Ход поиска и найденные варианты. Прогона не было — пустой ответ."""
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot)
    run = await analogs_service.latest_run(db, str(doc.task.id), doc.kind)
    return AnalogsStateResponse(**analogs_service.run_to_dict(run))


@router.post("/{card_id}/{kind}/analogs/cancel", response_model=AnalogsStateResponse)
async def cancel_analogs_search(
    card_id: str,
    kind: str,
    file_slot: Optional[str] = FileSlotQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Остановить идущий поиск. Уже найденные варианты остаются."""
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot)
    await analogs_service.cancel_run(db, str(doc.task.id), doc.kind)
    run = await analogs_service.latest_run(db, str(doc.task.id), doc.kind)
    return AnalogsStateResponse(**analogs_service.run_to_dict(run))


@router.get("/{card_id}/{kind}/history", response_model=list[HistoryEntryOut])
async def get_document_history(
    card_id: str,
    kind: str,
    file_slot: Optional[str] = FileSlotQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot)
    return [HistoryEntryOut(**entry) for entry in await svc.list_history(db, doc)]


@router.post("/{card_id}/{kind}/history/{entry_id}/revert", response_model=ApplyResponse)
async def revert_document(
    card_id: str,
    kind: str,
    entry_id: str,
    version_id: Optional[str] = Query(default=None),
    file_slot: Optional[str] = FileSlotQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Вернуть документ к состоянию до выбранной правки."""
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot)
    version = svc.pick_version(doc, version_id)
    result = await svc.revert_to_entry(db, doc, version, entry_id, current_user)
    return ApplyResponse(**result)


@router.post("/{card_id}/{kind}/heartbeat", response_model=HeartbeatResponse)
async def document_heartbeat(
    card_id: str,
    kind: str,
    file_slot: Optional[str] = FileSlotQuery,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Отметиться в документе и узнать, не редактирует ли его кто-то ещё."""
    doc = await svc.resolve_document(db, card_id, kind, current_user, file_slot)
    lock = await svc.heartbeat(db, str(doc.card.id), doc.kind, current_user)
    return HeartbeatResponse(lock=lock)
